src/schedule/zephyr_dma_domain.c

```diff
@@ -204,31 +249,58 @@ static int register_dma_irq(struct zephyr_dma_domain *domain,
 			if (core != crt_chan->core)
 				continue;
 
-			/* skip if IRQ was already registered */
-			if (crt_data->enabled_irq)
-				continue;
-
 			/* get IRQ number for current channel */
 			irq = interrupt_get_irq(dma_chan_irq(crt_dma, j),
 						dma_chan_irq_name(crt_dma, j));
 
-			crt_data->irq = irq;
-			crt_data->channel = crt_chan;
-			crt_data->dt = dt;
+			crt_irq_data = fetch_irq_data(domain, irq);
 
-			if (dt->started) {
-				/* if the Zephyr thread was started, we
-				 * can safely enable the DMA IRQs
+			if (!crt_irq_data) {
+				/* if there's no list item matching given IRQ then
+				 * that IRQ hasn't been allocated yet so we need to do
+				 * so here.
 				 */
-				ret = enable_dma_irq(crt_data);
-				if (ret < 0)
-					return ret;
+				crt_irq_data = rzalloc(SOF_MEM_ZONE_SYS_SHARED, 0,
+						SOF_MEM_CAPS_RAM, sizeof(*crt_irq_data));
+				if (!crt_irq_data)
+					return -ENOMEM;
+
+				list_init(&crt_irq_data->channels);
+
+				list_item_append(&crt_irq_data->list, &domain->irqs);
+				crt_irq_data->intid = irq;
+				crt_irq_data->dt = dt;
+			}
+
+			chan_data = rzalloc(SOF_MEM_ZONE_SYS_SHARED, 0,
+				SOF_MEM_CAPS_RAM, sizeof(*chan_data));
+			if (!chan_data)
+				return -ENOMEM;
+
+			irq_local_disable(flags);
+
+			list_item_append(&chan_data->list, &crt_irq_data->channels);
 
-				crt_data->enabled_irq = true;
+			chan_data->channel = crt_chan;
+			/* bind registrable ptask to channel */
+			chan_data->pipe_task = pipe_task;
+			chan_data->irq_data = crt_irq_data;
```

Comment:
in general it's better to first fully initialise an object and only then to add it to a global list

Reply:
ACK, will fix this.